doc/HttpTransportAdapter.md

```diff
@@ -0,0 +1,223 @@
+```
+
+## Building a Custom HTTP Transport Adapter
+
+The Azure SDK for C++ uses CMake options to define what HTTP transport adapters to build.
+
+You can see [CMake options](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#cmake-build-options) to learn about all the supported options and what is the specific option required to be set when building your own HTTP transport adapter.
+
+Follow these steps to implement your own HTTP transport adapter:
+
+1. Implement the interface.
+
+The HTTP transport adapter interface is define in `Azure::Core::HTTP::HttpTransport` from `inc/azure/core/http/transport.hpp`.
+
+```cpp
+/*
+* Implementing the HTTP transport adapter.
+*/
+
+#include "azure/core/http/transport.hpp"
+
+// Derive from HttpTransport
+class CustomTransportAdapter : public Azure::Core::Http::HttpTransport {
+  public:
+    // Override the Send method
+    std::unique_ptr<Azure::Core::Http::RawResponse> Send(
+        Azure::Core::Context const& context,
+        Azure::Core::Http::Request& request) override;
+};
+```
+
+> The implementation needs to take care of using the members from the `Request` argument which contains all the HTTP request information to write the HTTP request using some C++ HTTP stack. Then it needs to create the `RawResponse` from the response and return it. The `context` argument can be optionally used to check if the request is cancelled before performing I/O operations (like writing to the network).
+
+2. Stateless component.
+
+Make sure that the HTTP transport adapter that is created is stateless. This is because it can be shared from multiple SDK clients as a shared ptr. Consider designing a session object, for instance, if you need to hold some state. See the next pattern as an example.
+
+```cpp
+/*
+* Using a session pattern for the HTTP transport adapter.
+*/
+
+// Create a session class first.
+class CustomHttpTransportAdapterSession {
+  private:
+    // Keep any state like network/socket handlers here.
+    SomeHttpClientLibraryHandler handler;
+    SomeOtherRequiredState state;
+  
+  public:
+    // Expose any functionality required. For example, this could
+    // be the main method to produce the HTTP raw response.
+    std::unique_ptr<Azure::Core::Http::RawResponse> Perform(
+      Azure::Core::Context const& context,
+      Azure::Core::Http::Request& request
+    );
+}
+```
+
+Then, within the Send() method implementation.
+
+```cpp
+// Create a new Session object when calling Send().
+    std::unique_ptr<Azure::Core::Http::RawResponse> Send(
+        Azure::Core::Context const& context,
+        Azure::Core::Http::Request& request) {
+
+          // This will ensure that multiple clients can
+          // call Send() at the same time.
+          CustomHttpTransportAdapterSession session;
+
+          // Use any functionally from your created session.
+          // For this example, we could just return like this:
+          return session.Perform();
+        }
+
```

Comment:
Fix the indentation of this sample by removing the extra spacing in the beginning of each line.